File: 이것이 취업을 위한 코딩 테스트다/최단_경로/전보/전보.py
import heapq

INF = int(1e9)
n, m, c = map(int, input().split())
d = [INF] * (n + 1)
d[c] = 0
graph = [[] for _ in range(n + 1)]

for _ in range(m):
  x, y, z = map(int, input().split())
  graph[x].append((y, z))

q = []
heapq.heappush(q, (0, c))

while q:
  dist, now = heapq.heappop(q)

  if d[now] < dist:
    continue

  for (next, cost) in graph[now]:
    new_cost = dist + cost

    if new_cost < d[next]:
      d[next] = new_cost
      heapq.heappush(q, (new_cost, next))

count = 0
max_distance = 0

for i in d:
  if i != INF:
    count += 1
    max_distance = max(max_distance, i)

print(count - 1, max_distance)

File: 이것이 취업을 위한 코딩 테스트다/최단_경로/전보/전보.cpp
#include <iostream>
#include <vector>
#include <queue>
#define INF int(1e9)

using namespace std;


int main() {
  int d[30001];
  vector<pair<int, int>> graph[30001];
  priority_queue<pair<int, int>> q;
  int n, m, c, x, y, z;

  cin >> n >> m >> c;

  fill(d, d + (n + 1), INF);

  for (int i = 0; i < m ; i++) {
    cin >> x >> y >> z;
    graph[x].push_back({y, z});
  }

  q.push({0, c});
  d[c] = 0;

  while (!q.empty()) {
    int dist = -q.top().first;
    int now = q.top().second;
    q.pop();

    if (dist > d[now]) {
      continue;
    }

    for (int i = 0; i < graph[now].size(); i++) {
      int cost = graph[now][i].second;
      int next = graph[now][i].first;
      int new_cost = cost + dist;

      if (new_cost < d[next]) {
        d[next] = new_cost;

        q.push(make_pair(-new_cost, next));
      }
    }
  }
  
  int max_distance = 0, count = 0;

  for (int i = 1; i <= n; i++) {
    if (d[i] != INF) {
      count += 1;
      max_distance = max(max_distance, d[i]);
    }
  }

  cout << count - 1 << ' ' << max_distance;

  return 0;
}
